compiler/optimizations/gpuTransforms.cpp

```diff
@@ -606,8 +606,40 @@ void GpuKernel::generateEarlyReturn() {
   fn_->insertAtTail(new CondStmt(new SymExpr(isOOB), thenBlock));
 }
 
+static const std::unordered_map<PrimitiveTag, const char*>
+gpuPrimitivesDisallowedOnHost = {
+  { PRIM_GPU_BLOCKIDX_X, "getBlockIdxX" },
+  { PRIM_GPU_BLOCKIDX_Y, "getBlockIdxY" },
+  { PRIM_GPU_BLOCKIDX_Z, "getBlockIdxZ" },
+  { PRIM_GPU_BLOCKDIM_X, "getBlockDimX" },
+  { PRIM_GPU_BLOCKDIM_Y, "getBlockDimY" },
+  { PRIM_GPU_BLOCKDIM_Z, "getBlockDimZ" },
+  { PRIM_GPU_THREADIDX_X, "getThreadIdxX" },
+  { PRIM_GPU_THREADIDX_Y, "getThreadIdxY" },
+  { PRIM_GPU_THREADIDX_Z, "getThreadIdxZ" },
+  { PRIM_GPU_GRIDDIM_X, "getGridDimX" },
+  { PRIM_GPU_GRIDDIM_Y, "getGridDimY" },
+  { PRIM_GPU_GRIDDIM_Z, "getGridDimZ" },
+};
+
 void GpuKernel::populateBody(CForLoop *loop, FnSymbol *outlinedFunction) {
   std::set<Symbol*> handledSymbols;
+
+  std::vector<CallExpr*> callExprsInBody;
+  for_alist(node, loop->body) {
+    collectCallExprs(node, callExprsInBody);
+  }
+
+  for_vector(CallExpr, callExpr, callExprsInBody) {
+    if (callExpr->isPrimitive(PRIM_GPU_SET_BLOCKSIZE)) {
+      if (blockSize_ != nullptr) {
+        USR_FATAL(callExpr, "Can only set GPU block size once per GPU-eligible loop.");
+      }
+      blockSize_ = toSymExpr(callExpr->get(1));
+      callExpr->remove();
```

Comment:
Well, I think the node is already not copied into the outlined function, and is removed from the host loop (or a USR_FATAL is thrown), which is why I didn't add it to the map already. Also, currently, all the primitives are replaced with a "0" expression because they return ints, but the `setBlockSize` is void.